system/security/plugins/jwtSecurity/README.md

```diff
@@ -0,0 +1,164 @@
+## JWT Authorization Security Manager Plugin
+
+The purpose of this plugin is to provide authentication and authorization capabilities for HPCC Systems users, with the
+credentials passed via valid JWT tokens.
+
+The intention is to adhere as closely as possibly to the OpenID Connect (OIC) specification, which is a simple identity
+layer on top of the OAuth 2.0 protocol, while maintaining compatibility with the way HPCC Systems performs
+authentication and authorization today.  More information about the OpenID Connect specification can be found at
+<https://openid.net/specs/openid-connect-core-1_0.html>.
+
+One of the big advantages of OAuth 2.0 and OIC is that the service (in this case, HPCC Systems) never interacts with the
+user directly.  Instead, authentication is performed by a trusted third party and the (successful) results are passed to
+the service in the form of a verifiable encoded token.
+
+Unfortunately, HPCC Systems does not support the concept of third-party verification.  It assumes that users -- really,
+any client application that operates as a user, including things like IDEs -- will submit username/password credentials
+for authentication.  Until that is changed, HPCC Systems won't be able to fully adhere to the OIC specification.
+
+We can, however, implement *most* of the specification.  That is what this plugin does.
+
+### Code Documentation
+
+Doxygen (<https://www.doxygen.nl/index.html>) can be used to create nice HTML documentation for the code.  Call/caller
+graphs are also generated for functions if you have dot (https://www.graphviz.org/download/) installed and available on
+your path.
+
+Assuming ```doxygen``` is on your path, you can build the documentation via:
+
+	cd system/security/plugins/jwtSecurity
+	doxygen Doxyfile
+
+The documentation can then be accessed via ```docs/html/index.html```.
+
+### Theory of Operations
+
+The plugin is called by the HPCC Systems ```esp``` process when a user needs to be authenticated.  That call will
+contain the user's username and either a reference to a session token or a password.  The session token is present only
+for already-authenticated users.
+
+If the session token is not present, the plugin will call a ```JWT login service``` (also known as a JWT login endpoint)
+with the username and password, plus a nonce value for additional security.
+
+That service authenticates the username/password credentials.  If everything is good, the service constructs an
```

Comment:
Flavio may prefer it if we sent a hash of the password rather than the value.

Reply:
Good point.  I will work with the endpoint developer to get that implemented.

Reply:
Implemented SHA256.

Reply:
Should we consider PBKDF2, bcrypt or scrypt to make attacking the hash more difficult?  Or perhaps at least HMAC-SHA256?

Reply:
We only hash once per session and then use the token we receive so the overhead of doing this once may be worth it?

Reply:
HMAC requires a secret key.  Do you envision this as a configuration item?

bcrypt uses Blowfish as a hash, which is not one of Blowfish's original design goals.  OpenSSL doesn't support that use for that reason.

Reply:
I was just listing schemes to consider.  Don't think adding an extra key is desirable, unless one was already provided. Perhaps scrypt then, since it is included in openssl and I don't think has key requirements?

Reply:
I am not in favor of adding another secret, either.  If the goal is prevent rainbow attacks, what about SHA256(client_id + password)?  In other words, seeding the hash with the JWT client_id (which is defined as a configuration parameter) before hashing in the password?

Reply:
Actually, do it the other way around (pw + client_id).  That way the endpoint can store only sha256(pw) and then mix the client_id in to verify.

Reply:
Settled this a bit differently:  SHA256(SHA256(pw) + nonce) will be sent to the endpoint.  The endpoint saves SHA256(pw) in a database, so on that end they compare the sent hash with SHA256(retrieved_hash + nonce) to validate.